added homework files

Fozia-21 opened the pull request:
Added Javascript1 week4 homework

obelitz commented on javascript/javascript1/week4/voice-assistant.js:
```diff
@@ -0,0 +1,219 @@
+
+function getReply(command)
+{
+    let wordString = command.split(" "); 
+    let KeyWord = wordString[0];  // extract the keyword
```
These should be `const`.

obelitz replied:
Variable names are usually written in `camelCase`, so this would be `const keyWord`.

obelitz commented on javascript/javascript1/week4/voice-assistant.js:
```diff
@@ -0,0 +1,219 @@
+    let nameString;
+    let nameString2;
+
+    switch(KeyWord){
+        //--------------------------------------------------------------//
+        case 'Hello':
```
Good idea to check the first word!

obelitz replied:
If you want to make it to ignore case, you can 
```js
switch(keyWord.toLowerCase()) {
  case 'hello':
```

obelitz commented on javascript/javascript1/week4/voice-assistant.js:
```diff
@@ -0,0 +1,219 @@
+        //--------------------------------------------------------------//
+        case 'Hello':
+            nameString = wordString[4]; // extract the name (as it is on 4th index)
+            if(nameString == '')
```
Empty strings evaluate to false, so this could be written like:
`if(!nameString)`

obelitz commented on javascript/javascript1/week4/voice-assistant.js:
```diff
@@ -0,0 +1,219 @@
+                }
+                else
+                {
+                    console.log('name do not exists, so adding to the list');
+                    nameArray.push(nameString.toLowerCase()); // add to the array
+                    console.log(`Nice to meet you:  ${nameString} !`);
```
Oh, so you're doing a list of names. Ok 👍

obelitz commented on javascript/javascript1/week4/voice-assistant.js:
```diff
@@ -0,0 +1,219 @@
+                    console.log('name do not exists, so adding to the list');
+                    nameArray.push(nameString.toLowerCase()); // add to the array
+                    console.log(`Nice to meet you:  ${nameString} !`);
+                    //console.log(`Total Names in the array ${nameArray.length}`);  
```
I'd either uncomment this or delete it.

obelitz commented on javascript/javascript1/week4/voice-assistant.js:
```diff
@@ -0,0 +1,219 @@
+                
+                if(wordString.length > 4)
+                {
+                    nameString2 = wordString;
```
`nameString2` seems to be the wrong name for this value. 
You could maybe do `todoTask`

obelitz commented on javascript/javascript1/week4/voice-assistant.js:
```diff
@@ -0,0 +1,219 @@
+
+            case 'Add':    
+                
+                if(wordString.length > 4)
```
This is a good way to figure out if the task is longer than one word.

obelitz commented on javascript/javascript1/week4/voice-assistant.js:
```diff
@@ -0,0 +1,219 @@
+                {
+                    nameString2 = wordString;
+                    nameString2.splice(0, 1); // remove first element
+                    nameString2.splice(-1, 1); // remove last element
+                    nameString2.splice(-1, 1); // remove 2nd last element, whih is infact last now
```
Or you can do:
`todoWords.splice(-2)` which will delete all elements from the second to last element.

obelitz replied:
Or you can do:
`todoTask.splice(-2)` which will delete all elements from the second to last element.

obelitz commented on javascript/javascript1/week4/voice-assistant.js:
```diff
@@ -0,0 +1,219 @@
+                 // check if the item is in the list, before removing it.If exists then remove it.
+                if(todoArray.includes(nameString2.toLowerCase()))
+                {
+                    itemIndex = todoArray.indexOf(nameString2);
```
Remember to assign this before using. This is a `const`.

obelitz commented on javascript/javascript1/week4/voice-assistant.js:
```diff
@@ -0,0 +1,219 @@
+            
+            //----------------------------------
+            case 'What':    
+                nameString2 = wordString;
```
Still shouldn't be `nameString2`. 
You could just assign a new variable right here: 
`let nameCommand = wordString`

obelitz commented on javascript/javascript1/week4/voice-assistant.js:
```diff
@@ -0,0 +1,219 @@
+
+                if(nameString2 == 'is my name?')
+                {
+                    if(nameArray == ' ') 
```
Since nameArray is an array, I'd check the length:
`if(nameArray.length) {`

obelitz commented on javascript/javascript1/week4/voice-assistant.js:
```diff
@@ -0,0 +1,219 @@
+                else
+                {
+                    // mathematic operations
+                    let operator = wordString.slice(-2,-1);  // get the maths operator
+                    let operandA = parseInt(wordString.slice(-3,-2));
+                    let operandB = parseInt(wordString.slice(-1));
```
If you want, you can use what is called "array destructuring" here:
```js
// if wordString is ["what", "is", "2", "+", "2"]
const [ firstElement, secondElement, operandA, operator, operandB ] = wordString; 
// this will assign the values directly.
// and because you don't need the two first elements you can simply do:
const [ , , operandA, operator, operandB ] = wordString; 
```

obelitz commented on javascript/javascript1/week4/voice-assistant.js:
```diff
@@ -0,0 +1,219 @@
+                    console.log('Waiting for time out ....'); 
+                    
+                    let timeout;
+                    if((units='seconds') || (units='sec') )
```
This won't work.
If you want to check a value you need to use at least `==`. Ideally `===`.

obelitz commented on javascript/javascript1/week4/voice-assistant.js:
```diff
@@ -0,0 +1,219 @@
+
+            //--------------------------------------------------------------
+            default:
+                console.log('invalid command'); 
```
Good fallback. Potentially add `console.log('invalid command. try asking me "what day is it"')`

obelitz reviewed (commented):
Good job!

Notes:
- I have an idea on how you can avoid all the `.splice`.
- You can assign more variables. Don't use nameString2 for everything. The name is not descriptive enough.

Fozia-21 commented:
Thank you so much .
I implemented all suggestions 🙂

Sent from my iPhone

> On 1 Jun 2021, at 11.29 PM, obelitz ***@***.***> wrote:
> 
> ﻿
> @obelitz commented on this pull request.
> 
> Good job!
> 
> Notes:
> 
> I have an idea on how you can avoid all the .splice.
> You can assign more variables. Don't use nameString2 for everything. The name is not descriptive enough.
> In javascript/javascript1/week4/voice-assistant.js:
> 
> > +    let wordString = command.split(" "); 
> +    let KeyWord = wordString[0];  // extract the keyword
> These should be const.
> 
> In javascript/javascript1/week4/voice-assistant.js:
> 
> > +    switch(KeyWord){
> +        //--------------------------------------------------------------//
> +        case 'Hello':
> Good idea to check the first word!
> 
> In javascript/javascript1/week4/voice-assistant.js:
> 
> > +
> +let nameArray = [];
> +let todoArray = [];
> +
> +function getReply(command)
> +{
> +    let wordString = command.split(" "); 
> +    let KeyWord = wordString[0];  // extract the keyword
> +    let nameString;
> +    let nameString2;
> +
> +    switch(KeyWord){
> +        //--------------------------------------------------------------//
> +        case 'Hello':
> +            nameString = wordString[4]; // extract the name (as it is on 4th index)
> +            if(nameString == '')
> Empty strings evaluate to false, so this could be written like:
> if(!nameString)
> 
> In javascript/javascript1/week4/voice-assistant.js:
> 
> > +                    console.log('name do not exists, so adding to the list');
> +                    nameArray.push(nameString.toLowerCase()); // add to the array
> +                    console.log(`Nice to meet you:  ${nameString} !`);
> Oh, so you're doing a list of names. Ok 👍
> 
> In javascript/javascript1/week4/voice-assistant.js:
> 
> > +            {
> +                console.log('please enter your name');
> +            }
> +            else
> +            {    
> +                // compare with name array and add if not exist or skip if its already exist
> +                if(nameArray.includes(nameString.toLowerCase()) )
> +                {
> +                    console.log('Name already exists, so skipping it');
> +                }
> +                else
> +                {
> +                    console.log('name do not exists, so adding to the list');
> +                    nameArray.push(nameString.toLowerCase()); // add to the array
> +                    console.log(`Nice to meet you:  ${nameString} !`);
> +                    //console.log(`Total Names in the array ${nameArray.length}`);  
> I'd either uncomment this or delete it.
> 
> In javascript/javascript1/week4/voice-assistant.js:
> 
> > +                    console.log('name do not exists, so adding to the list');
> +                    nameArray.push(nameString.toLowerCase()); // add to the array
> +                    console.log(`Nice to meet you:  ${nameString} !`);
> +                    //console.log(`Total Names in the array ${nameArray.length}`);  
> +                }
> +        
> +            }
> +            break;
> +
> +        //--------------------------------------------------------------
> +
> +            case 'Add':    
> +                
> +                if(wordString.length > 4)
> +                {
> +                    nameString2 = wordString;
> nameString2 seems to be the wrong name for this value.
> You could maybe do todoTask
> 
> In javascript/javascript1/week4/voice-assistant.js:
> 
> > +                else
> +                {
> +                    console.log('name do not exists, so adding to the list');
> +                    nameArray.push(nameString.toLowerCase()); // add to the array
> +                    console.log(`Nice to meet you:  ${nameString} !`);
> +                    //console.log(`Total Names in the array ${nameArray.length}`);  
> +                }
> +        
> +            }
> +            break;
> +
> +        //--------------------------------------------------------------
> +
> +            case 'Add':    
> +                
> +                if(wordString.length > 4)
> This is a good way to figure out if the task is longer than one word.
> 
> In javascript/javascript1/week4/voice-assistant.js:
> 
> > +                    nameString2.splice(-1, 1); // remove last element
> +                    nameString2.splice(-1, 1); // remove 2nd last element, whih is infact last now
> Or you can do:
> todoWords.splice(-2) which will delete all elements from the second to last element.
> 
> In javascript/javascript1/week4/voice-assistant.js:
> 
> > +                    nameString2.splice(-1, 1); // remove last element
> +                    nameString2.splice(-1, 1); // remove 2nd last element, whih is infact last now
> Or you can do:
> todoTask.splice(-2) which will delete all elements from the second to last element.
> 
> In javascript/javascript1/week4/voice-assistant.js:
> 
> > +                {
> +                    nameString2 = wordString;
> +                    nameString2.splice(0, 1); // remove first element
> +                    nameString2.splice(-1, 1); // remove last element
> +                    nameString2.splice(-1, 1); // remove 2nd last element, whih is infact the last now
> +                    nameString2 = nameString2.join(' '); // combine element to form the string
> +                }
> +                else
> +                {
> +                    nameString2 = wordString[1];
> +                }
> +
> +                 // check if the item is in the list, before removing it.If exists then remove it.
> +                if(todoArray.includes(nameString2.toLowerCase()))
> +                {
> +                    itemIndex = todoArray.indexOf(nameString2);
> Remember to assign this before using. This is a const.
> 
> In javascript/javascript1/week4/voice-assistant.js:
> 
> > +                    itemIndex = todoArray.indexOf(nameString2);
> +                    todoArray.splice(itemIndex, 1) ; 
> +                    console.log(`${nameString2} is removed from myToDo`);              
> +                }
> +                else 
> +                {    
> +                    // if not in the list, print invalid item to be removed
> +                    // if item exist in the list, remove it :-) 
> +                    console.log(`Invalid item to be removed`);
> +                }
> +                break;
> +
> +            
> +            //----------------------------------
> +            case 'What':    
> +                nameString2 = wordString;
> Still shouldn't be nameString2.
> You could just assign a new variable right here:
> let nameCommand = wordString
> 
> In javascript/javascript1/week4/voice-assistant.js:
> 
> > +                    // if not in the list, print invalid item to be removed
> +                    // if item exist in the list, remove it :-) 
> +                    console.log(`Invalid item to be removed`);
> +                }
> +                break;
> +
> +            
> +            //----------------------------------
> +            case 'What':    
> +                nameString2 = wordString;
> +                nameString2.splice(0, 1); // remove first element
> +                nameString2 = nameString2.join(' '); // combine element to form the string
> +
> +                if(nameString2 == 'is my name?')
> +                {
> +                    if(nameArray == ' ') 
> Since nameArray is an array, I'd check the length:
> if(nameArray.length) {
> 
> In javascript/javascript1/week4/voice-assistant.js:
> 
> > +                    let operator = wordString.slice(-2,-1);  // get the maths operator
> +                    let operandA = parseInt(wordString.slice(-3,-2));
> +                    let operandB = parseInt(wordString.slice(-1));
> If you want, you can use what is called "array destructuring" here:
> 
> // if wordString is ["what", "is", "2", "+", "2"]
> const [ firstElement, secondElement, operandA, operator, operandB ] = wordString; 
> // this will assign the values directly.
> // and because you don't need the two first elements you can simply do:
> const [ , , operandA, operator, operandB ] = wordString; 
> In javascript/javascript1/week4/voice-assistant.js:
> 
> > +                break;
> +
> +            //-----------------------------------------
> +            case 'Set':  
> +                nameString2 = wordString.slice(1, 3); // get keyword like 'a timer'
> +                nameString2 = nameString2.join(' '); // combine element to form the string
> +                
> +                if(nameString2 == 'a timer')
> +                {
> +                    let units = wordString.slice(-1); 
> +                    let timervalue = wordString.slice(-2,-1);  
> +                    console.log(`Timer set for ${timervalue} ${units}`);  
> +                    console.log('Waiting for time out ....'); 
> +                    
> +                    let timeout;
> +                    if((units='seconds') || (units='sec') )
> This won't work.
> If you want to check a value you need to use at least ==. Ideally ===.
> 
> In javascript/javascript1/week4/voice-assistant.js:
> 
> > +                    }    
> +                    else if((units = 'minutes') || (units = 'mins') )
> +                    {
> +                        timeout = timervalue * 1000 * 60;  // represented a milliseconds
> +                    }
> +
> +                    setTimeout(function(){ console.log("Timer done"); }, timeout);
> +
> +                }
> +                break;
> +
> +
> +
> +            //--------------------------------------------------------------
> +            default:
> +                console.log('invalid command'); 
> Good fallback. Potentially add console.log('invalid command. try asking me "what day is it"')
> 
> In javascript/javascript1/week4/voice-assistant.js:
> 
> > +    let wordString = command.split(" "); 
> +    let KeyWord = wordString[0];  // extract the keyword
> Variable names are usually written in camelCase, so this would be const keyWord.
> 
> In javascript/javascript1/week4/voice-assistant.js:
> 
> > +    switch(KeyWord){
> +        //--------------------------------------------------------------//
> +        case 'Hello':
> If you want to make it to ignore case, you can
> 
> switch(keyWord.toLowerCase()) {
>   case 'hello':
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or unsubscribe.
